barretenberg/cpp/src/barretenberg/translator_vm/goblin_translator_flavor.hpp

```diff
@@ -732,53 +749,89 @@ class GoblinTranslatorFlavor {
         return builder.get_circuit_subgroup_size(compute_total_num_gates(builder));
     }
 
+    /**
+     * @brief A field element for each entity of the flavor.  These entities represent the prover polynomials
+     * evaluated at one point.
+     */
+    class AllValues : public AllEntities<FF> {
+      public:
+        using Base = AllEntities<FF>;
+        using Base::Base;
+    };
+    /**
+     * @brief A container for the prover polynomials handles.
+     */
+    class ProverPolynomials : public AllEntities<Polynomial> {
+      public:
+        // Define all operations as default, except copy construction/assignment
+        ProverPolynomials() = default;
+        // Constructor to init all unshifted polys to the zero polynomial and set the shifted poly data
+        ProverPolynomials(size_t circuit_size)
+        {
+            for (auto& poly : get_unshifted()) {
+                poly = Polynomial{ circuit_size };
+            }
+            set_shifted();
+        }
+        ProverPolynomials& operator=(const ProverPolynomials&) = delete;
+        ProverPolynomials(const ProverPolynomials& o) = delete;
+        ProverPolynomials(ProverPolynomials&& o) noexcept = default;
+        ProverPolynomials& operator=(ProverPolynomials&& o) noexcept = default;
+        ~ProverPolynomials() = default;
+        [[nodiscard]] size_t get_polynomial_size() const { return this->op.size(); }
+        /**
+         * @brief Returns the evaluations of all prover polynomials at one point on the boolean
+         * hypercube, which represents one row in the execution trace.
+         */
+        [[nodiscard]] AllValues get_row(size_t row_idx) const
+        {
+            AllValues result;
+            for (auto [result_field, polynomial] : zip_view(result.get_all(), this->get_all())) {
+                result_field = polynomial[row_idx];
+            }
+            return result;
+        }
+        // Set all shifted polynomials based on their to-be-shifted counterpart
+        void set_shifted()
+        {
+            for (auto [shifted, to_be_shifted] : zip_view(get_shifted(), get_to_be_shifted())) {
+                shifted = to_be_shifted.shifted();
+            }
+        }
+    };
+
     /**
      * @brief The proving key is responsible for storing the polynomials used by the prover.
-     * @note TODO(Cody): Maybe multiple inheritance is the right thing here. In that case, nothing should eve
-     * inherit from ProvingKey.
+     *
      */
-    class ProvingKey : public ProvingKey_<PrecomputedEntities<Polynomial>, WitnessEntities<Polynomial>, CommitmentKey> {
```

Comment:
Worth noting that this inheritance let us reuse a single getter definition, where now we have loose coupling. Maybe issue for some static checks or tests?

Reply:
As we discussed offline, this is handled by ProverPolynomials being an AllEntities which is constructed via inheritance from the sub components. (True for all flavors except Ultra for some reason. Made TODO for this)